### ISSUE TITLE
n_cleanup_primes() is not thread safe

### ISSUE THREAD

**Issue by peter-frentrup** [wbhart/flint2#541]
when FLINT_REENTRANT is defined but HAVE_TLS is not, we use pthread_mutex_lock(&primes_lock) inside n_compute_primes() but not in n_cleanup_primes() which will cause a crash when calling flint_cleanup() from multiple threads. The fix would be to add

> #if FLINT_REENTRANT && !HAVE_TLS
> pthread_mutex_lock(&primes_lock);
> #endif
> //...
> #if FLINT_REENTRANT && !HAVE_TLS
> pthread_mutex_unlock(&primes_lock);
> #endif

to n_cleanup_primes(). However, primes_lock is only locally defined in compute_primes.c. I am not sure whether is is a good idea to put a `pthread_mutex_t primes_lock;` declaration into the public header "ulong_extras.h". The possibilities to fix this are:

1) Put `extern pthread_mutex_t primes_lock;` into cleanup_primes.c and add the relevant locking there.
2) Declare the primes_lock variable in "ulong_extras.h" and fix n_cleanup_primes() in cleanup_primes.c. But then the lock would be publicly visible (if FLINT_REENTRANT && !HAVE_TLS).
3) Introduce a new header "private_ulong_extras.h" with the necessary declaration and fix n_cleanup_primes() in cleanup_primes.c.
4) Move the corrected n_cleanup_primes() into compute_primes.c and delete the file cleanup_primes.c completely.

I favor option 4) but it goes against the perceived rule "one file = one public function". Which one should I implement? Also, should there be a `pthread_once(&primes_initialised, n_compute_primes_init);` in n_cleanup_primes() or is it considered a bug to call n_cleanup_primes() before n_compute_primes()?

Best regards,
Peter Frentrup

**Comment by wbhart** [comment]
n_compute_primes has to be called multiple times if you want to compute more and more primes. But it should be a bug to call n_cleanup_primes() from anything other than the master thread. It's a global object which Flint assumes exists once it is first called.

Can you think of a compelling argument for why we would want to allow that?

**Comment by peter-frentrup** [comment]
The library I work on (part of https://github.com/peter-frentrup/pMath) has no concept of a main thread. Every thread it uses has a cleanup function. 

The documentation also suggests that n_cleanup_primes() can or should be called from every thread that uses n_compute_primes() etc. and in fact n_compute_primes() calls flint_register_cleanup_function(n_cleanup_primes). 

So n_cleanup_primes() will be called by FLINT via flint_cleanup(), which should be called on every thread before it exits. I am not calling n_cleanup_primes() myself, only flint_cleanup() on each thread.

My current workaround is to ensure that FLINT_REENTRANT and HAVE_TLS are both defined as 1.

**Comment by wbhart** [comment]
My mistake, I think there is in fact a primes array per thread, as we use thread local storage. 

On the other hand, there should be no problem with multiple threads cleaning up the same primes array in that case. I'll have to discuss this with my colleagues, as it is not clear what we actually intended.

Do @tthsqe12 or @fredrik-johansson have any input on this?

**Comment by wbhart** [comment]
Actually, I think think the mutex is only used if you don't have thread local storage. I see the problem now. Yes, we need to only clean up once in that case.

**Comment by wbhart** [comment]
Except that this won't work if one thread cleans up before another has finished with the primes.

**Comment by wbhart** [comment]
I've taken a careful look at this and the current implementation is incorrect when TLS is not available. The first thread to compute some primes registers a cleanup function. But this is called by the first thread to call the flint_cleanup function, not the last.

@tthsqe12 's objection is not actually a problem because the only function that can actually alter the primes array is n_compute_primes, which is guarded. Moreover, the structure it creates is never actually altered, it's just duplicated and an index to the new structure is updated once it exists. Any threads that try to read from the old structure will be fine, and any that try to read from the new structure are fine.

But this doesn't solve @peter-frentrup 's problem. Unfortunately, there is no clean solution without TLS. So I offer the following possibilities:

1) Make it so that no cleanup function is registered when TLS is not available, and just leak.
2) Register the cleanup function, but only call cleanup functions (when no TLS is available) when an explicit call to flint_cleanup_master is called, so you have control over when it happens.
3) Add a new function, n_primes_init_first() which has to be called precisely once (when no TLS is available) from each thread before any subsequent calls to n_primes_init(). Everything would be automatic, so long as all threads call flint_cleanup when exiting (it would just use a counter and only actually clean up when the counter reaches 0).

Although a completely automatic implementation is possible, it's would be a major work, and would require making a hacky, cross-platform TLS solution to make it work, or it would be extremely expensive at runtime, e.g. keeping a hash table of tid's.

@peter-frentrup Can you 1) let me know what platform you have this problem on. Why is TLS not available? 2) Let me know which of the above solutions you prefer. I would need your help to test it, as I really don't know of any systems without TLS nowadays that I have access to, which explains why people are not beating our door down over this.

**Comment by peter-frentrup** [comment]
I only stumbled accidentally over this problem because I built FLINT myself using CMAKE+MSVC on Windows 7 (32 and 64 bit). I forgot the MSVC version, maybe 2015 or 2017. It seems that HAVE_TLS was undefined or set to 0 by default in this setup. The proper setting for my system is indeed to set HAVE_TLS=1.

Hence, the solution has no direct impact on me. Nevertheless, I prefer solution 2. The problem with 1 is that leaking memory is very annoying for people who use valgrind or similar tools. The problem with 3 is that is breaks existing code, because nobody calls n_primes_init_first() at the moment.
